pkg/tls/certificatemanagement/ca.go

```diff
@@ -0,0 +1,596 @@
+			return nil, fmt.Errorf("unable to create secret: %s", err)
+		}
+		kp.Certificate, err = parseCertificate(secret.Data[corev1.TLSCertKey])
+		if err != nil {
+			return nil, err
+		}
+	}
+	kp.secret = secret
+	return kp, nil
+}
+
+// Issued returns true if the provided certificate was signed by certificateManager and returns false for user provided KeyPairs or Certificate Management.
+func (cm *certificateManager) Issued(cert Certificate) bool {
+	return cm.CertificateManagement() == nil && cert != nil &&
+		string(cert.X509Certificate().AuthorityKeyId) == string((cm.X509Certificate()).AuthorityKeyId)
+}
+
+// GetCertificate returns a Certificate. If the certificate is not found a k8s.io NotFound error is returned.
+func (cm *certificateManager) GetCertificate(cli client.Client, secretName, secretNamespace string) (Certificate, error) {
+	if cm.CertificateManagement() != nil {
+		return certificateManagementKeyPair(cm, secretName, secretNamespace, nil), nil
+	}
+	secret := &corev1.Secret{}
+	err := cli.Get(context.Background(), types.NamespacedName{
+		Name:      secretName,
+		Namespace: secretNamespace,
+	}, secret)
+	if err != nil {
+		return nil, err
+	}
+	err = standardizeFields(secret)
+	if err != nil {
+		return nil, err
+	}
+	x509Cert, err := parseCertificate(secret.Data[corev1.TLSCertKey])
+	if err != nil {
+		return nil, err
+	}
+	if x509Cert.NotAfter.Before(time.Now()) || x509Cert.NotBefore.After(time.Now()) {
+		return nil, fmt.Errorf("secret %s is not valid at this date", secretName)
+	}
+	return &keyPair{ca: cm, Certificate: x509Cert, secret: secret}, nil
+}
+
+// GetKeyPair returns an existing KeyPair. If the KeyPair is not found, a k8s.io NotFound error is returned.
+func (cm *certificateManager) GetKeyPair(cli client.Client, secretName, secretNamespace string) (KeyPair, error) {
+	secret := &corev1.Secret{}
+	err := cli.Get(context.Background(), types.NamespacedName{
+		Name:      secretName,
+		Namespace: secretNamespace,
+	}, secret)
+	if err != nil {
+		if !kerrors.IsNotFound(err) && cm.keyPair.CertificateManagement != nil {
+			return certificateManagementKeyPair(cm, secretName, secretNamespace, nil), nil
+		}
+		return nil, nil
+	}
+	err = standardizeFields(secret)
+	if err != nil {
+		return nil, err
+	}
+	certificate, err := parseCertificate(secret.Data[corev1.TLSCertKey])
+	if err != nil {
+		return nil, err
+	}
+	if certificate.NotAfter.Before(time.Now()) || certificate.NotBefore.After(time.Now()) {
+		return nil, fmt.Errorf("secret %s is not valid at this date", secretName)
+	}
+	return &keyPair{
+		ca:          cm,
+		secret:      secret,
+		Certificate: certificate,
+	}, nil
+}
+
+// CertificateManagement returns the CertificateManagement object or nil if it is not configured.
+func (cm *certificateManager) CertificateManagement() *operatorv1.CertificateManagement {
+	return cm.keyPair.CertificateManagement
+}
+
+type keyPair struct {
+	csrImage string
+	*x509.Certificate
+	secret        *corev1.Secret
```

Comment:
Is there any benefit to actually storing the secret as opposed to just storing the name and generating the secret fresh when you call `Secret(namespace)`? You're already storing the certificate as a field so you're duplicating that.